lib/metasploit/framework/compiler/golang.rb

```diff
@@ -0,0 +1,84 @@
+require 'open3'
+require 'fileutils'
+
+module Metasploit
+  module Framework
+    module Compiler
+      class Golang
+
+        UncompilablePayloadError     = Class.new(StandardError)
+        CompiledPayloadNotFoundError = Class.new(StandardError)
+        GoModInitError               = Class.new(StandardError)
+
+        DefaultBuilder = "go"
+
+        attr_reader :outfile, :env, :ldflags, :builder, :keep_src, :builder_flags, :build_flags
+        def initialize(**opts)
+          @env = opts[:env]
+          @ldflags  = opts[:ldflags]
+          @builder_flags  = opts[:builder_flags]
+          @build_flags  = opts[:build_flags]
+          @outfile  = opts[:outfile]
+          @keep_src = opts[:keep_src]
+          @builder  = opts[:builder || DefaultBuilder]
+        end
+
+        def self.available?
+            !!(Msf::Util::Helper.which("go"))
+        end
+
+        def go_build_src(src)
+          Dir.mktmpdir do |dir|
+            Dir.chdir(dir) do
+              Tempfile.create(['main_', ".go"], dir) do |file|
+                File.write(file, src)
+                mod_init
+                cmd = cmd_build.push(file)
+                if keep_src
+                  FileUtils.cp(file, File.dirname(outfile))
+                end
+                exec(cmd)
+              end
+            end
+          end
+        end
+
+        def go_build_files(files)
+          Dir.mktmpdir do |dir|
+            Dir.chdir(dir) do
+              files.each { |f| FileUtils.cp(f, dir) }
+              mod_init
+              exec(cmd_build)
+            end
+          end
+        end
+
+        def go_build(file)
+          go_build_files([file])
+        end
+
+        def cmd_build
+          cmd = [builder]
+          cmd << "build #{build_flags}"
+          cmd << "-o #{outfile}"
+          cmd << "-ldflags '#{ldflags}'" if ldflags
+        end
+
+        def exec(cmd)
+          stdout_err, status = Open3.capture2e(env, cmd.join(" "))
```

Comment:
Should directly use the argument array here. That would eliminate any problems with bad chars that could be in `outfile` for example.

```suggestion
          stdout_err, status = Open3.capture2e(env, *cmd)
```

You'll need to update `cmd_build` to treat each flag and value as it's own member in the array though.

Reply:
Hmm, ok. I had it that way at first, but given the variadic nature of build args and ldflags, depending on the preference of the user, this seemed to be the safer way. I can make it work like you're suggesting by joining `cmd`, then splitting again on `<space>` but that seems strange to me. What bad characters could one enter on their keyboard that would cause an error?

Reply:
missed your note about changing cmd_build. i'll see what i can do